core/src/stored/append.cc

```diff
@@ -111,6 +118,164 @@ static bool SaveFullyProcessedFilesAttributes(
+      bytes_out += content->size;
+    }
+    element_removed.notify_one();
+    return t;
+  }
+
+  BareosSocket* close_and_get()
+  {
+    end.store(true);
+    receive_thread.join();
+    return fd;
+  }
+
+ private:
+  bool error_while_reading{false};
+  std::atomic<bool> end{false};
+
+  BareosSocket* fd;
+  // channel::out<result_type> chan;
+
+  std::mutex mut{};
+  bool finished{false};
+  std::size_t bytes_out{0};
+  std::condition_variable not_empty{};
+  std::condition_variable element_removed{};
+  std::deque<result_type> results{};
+
+  std::deque<result_type> cache;
+
+  // this has to be defined last!
+  std::thread receive_thread;
+  void do_work()
+  {
+    POOLMEM* save = fd->msg;
+    bool cont = true;
+    std::size_t max_size{0};
+    std::size_t max_bytes{0};
+    std::size_t bytes_in{0};
+    std::size_t message_count{0};
+    std::size_t full_message_count{0};
+    std::size_t signal_count{0};
+    for (int res = 0; cont; res = fd->WaitData(0, 100'000)) {
+      if (res == 1) {
+        PoolMem msg(PM_MESSAGE);
+        fd->msg = msg.addr();
+        result_type result;
+        int n = BgetMsg(fd);
+        // fd->msg might have been relocated
+        msg.addr() = fd->msg;
+        if (n < 0) {
+          if (n == BNET_SIGNAL) {
+            result = signal{fd->message_length};
+            // break; /* end of data */
+          } else if (n == BNET_HARDEOF) {
+            result = error{error::type::HARDEOF, fd->bstrerror()};
+            cont = false;
+          } else {
+            result = error{error::type::INTERNAL_ERROR, fd->bstrerror()};
+            cont = false;
+          }
+
+          signal_count += 1;
+        } else {
+          std::size_t length = n;
+          result = message{length, std::move(msg)};
+          if (n == 64 * 1024) full_message_count += 1;
+          message_count += 1;
+        }
+        fd->msg = nullptr;
+
+        {
+          std::unique_lock l(mut);
+          if (n > 0) { bytes_in += n; }
+          bytes_in += sizeof(result);
+
+          element_removed.wait(l, [this, bytes_in] {
+            return bytes_in - bytes_out < 5ULL * 1024ULL * 1024ULL
+                   || end.load();
+          });
+
+          results.emplace_back(std::move(result));
+          if (results.size() > max_size) { max_size = results.size(); }
+          ASSERT(bytes_in >= full_message_count * 64ULL * 1024ULL);
+          if (bytes_in - bytes_out > max_bytes) {
+            ASSERT(bytes_in > bytes_out);
+            max_bytes = bytes_in - bytes_out;
+          }
+          if (message_count % 10000 == 0) {
+            Dmsg1(50,
+                  "MsgQueue Message Count = %llu, Signal Count = %llu, Size = "
+                  "%llu\n",
+                  message_count, signal_count, bytes_in - bytes_out);
+          }
+          not_empty.notify_one();
+        }
+      } else if (res == -1) {
+        cont = false;
+        error_while_reading = true;
+      }
```

Comment:
will `fd->WaitData()` only return `1` or `-1`?

Reply:
maybe `WaitData()` should return an enum instead ? What do you think ?